cpp/community_detection_module/algorithm/louvain.cpp

```diff
@@ -0,0 +1,145 @@
+#include "louvain.hpp"
+
+namespace louvain_alg {
+
+namespace {
+
+constexpr int kReplaceMap = 0;
+constexpr int kThreadsOpt = 1;
+constexpr int kNumColors = 16;
+
+std::vector<std::int64_t> GrappoloCommunityDetection(GrappoloGraph &grappolo_graph, bool coloring,
+                                                     std::uint64_t min_graph_size, double threshold,
+                                                     double coloring_threshold) {
+  auto number_of_vertices = grappolo_graph.numVertices;
+
+  auto *cluster_array = (long *)malloc(number_of_vertices * sizeof(long));
+#pragma omp parallel for
+  for (long i = 0; i < number_of_vertices; i++) {
+    cluster_array[i] = -1;
+  }
+
+  // Dynamically set currently.
+  auto num_threads = omp_get_num_threads();
+  if (coloring) {
+    runMultiPhaseColoring(&grappolo_graph, cluster_array, coloring, kNumColors, kReplaceMap, min_graph_size, threshold,
+                          coloring_threshold, num_threads, kThreadsOpt);
+  } else {
+    runMultiPhaseBasic(&grappolo_graph, cluster_array, kReplaceMap, min_graph_size, threshold, coloring_threshold,
+                       num_threads, kThreadsOpt);
+  }
+
+  // Store clustering information in vector
+  std::vector<std::int64_t> result;
+  for (long i = 0; i < number_of_vertices; ++i) {
+    result.emplace_back(cluster_array[i]);
+  }
+  // Detach memory, no need to free graph instance, it will be removed inside algorithm
+  free(cluster_array);
+
+  // Return empty vector if algorithm has failed
+  return result;
+}
+
+void LoadUndirectedEdges(const mg_graph::GraphView<> &memgraph_graph, GrappoloGraph &grappolo_graph) {
+  int num_threads = 1;
+#pragma omp parallel
+  { num_threads = omp_get_num_threads(); }
+
+  auto number_of_vertices = memgraph_graph.Nodes().size();
+  auto number_of_edges = memgraph_graph.Edges().size();
+
+  // Case without graph edges
+  if (number_of_edges == 0) {
+    grappolo_graph.numEdges = 0;
+    return;
+  }
+
+  auto tmp_edge_list = std::unique_ptr<edge[]>(new edge[number_of_edges]);  // Every edge stored ONCE
+
+  // TODO: (jmatak) Add different weights on edges
+  double default_weight = 1.0;
+  std::uint64_t edge_index = 0;
+  for (const auto [id, from, to] : memgraph_graph.Edges()) {
+    tmp_edge_list[edge_index].head = from;              // The S index
+    tmp_edge_list[edge_index].tail = to;                // The T index: Zero-based indexing
+    tmp_edge_list[edge_index].weight = default_weight;  // Make it positive and cast to Double, fixed to 1.0
+    edge_index++;
+  }
```

Comment:
Why can't `edge_index` just be `memgraph_graph.Edges().size()` ?

Reply:
Because `edge_index` is there for the temporary list, and `memgraph_graph.Edges().size()` is constant.

Reply:
Aha, sorry I got confused